src/buildblock/ML_norm.cxx

```diff
@@ -129,68 +129,135 @@ void make_det_pair_data(DetPairData& det_pair_data,
-  {
-    fan_indices[a] = 
-      IndexRange<1>(a+num_detectors/2-half_fan_size,
-                    a+num_detectors/2+half_fan_size);
-  }
-  det_pair_data.grow(fan_indices);
-  det_pair_data.fill(0);
+  if(proj_data_info_general_type.get_scanner_ptr()->get_scanner_geometry()== "BlocksOnCylindrical")
+    {
+      const ProjDataInfoBlocksOnCylindricalNoArcCorr& proj_data_info =
+        dynamic_cast<const ProjDataInfoBlocksOnCylindricalNoArcCorr&>(proj_data_info_general_type);
+
+      const int num_detectors =
+        proj_data_info.get_scanner_ptr()->get_num_detectors_per_ring();
+      const int fan_size =
+        2*max(proj_data_info.get_max_tangential_pos_num(),
+              -proj_data_info.get_min_tangential_pos_num()) + 1;
+      // fan will range from -half_fan_size to +half_fan_size (i.e. an odd number of elements)
+      const int half_fan_size = fan_size/2;
+
+      IndexRange<2> fan_indices;
+      fan_indices.grow(0,num_detectors-1);
+      for (int a = 0; a < num_detectors; ++a)
+        {
+          fan_indices[a] =
+            IndexRange<1>(a+num_detectors/2-half_fan_size,
+                          a+num_detectors/2+half_fan_size);
+        }
+      det_pair_data.grow(fan_indices);
+      det_pair_data.fill(0);      
+    }
+  else
+    {
+      const ProjDataInfoCylindricalNoArcCorr& proj_data_info =
+	dynamic_cast<const ProjDataInfoCylindricalNoArcCorr&>(proj_data_info_general_type);
+
+      const int num_detectors = 
+	proj_data_info.get_scanner_ptr()->get_num_detectors_per_ring();
+      const int fan_size = 
+	2*max(proj_data_info.get_max_tangential_pos_num(),
+	      -proj_data_info.get_min_tangential_pos_num()) + 1;
+      // fan will range from -half_fan_size to +half_fan_size (i.e. an odd number of elements)
+      const int half_fan_size = fan_size/2;
+
+      IndexRange<2> fan_indices;
+      fan_indices.grow(0,num_detectors-1);
+      for (int a = 0; a < num_detectors; ++a)
+	{
+	  fan_indices[a] = 
+	    IndexRange<1>(a+num_detectors/2-half_fan_size,
+			  a+num_detectors/2+half_fan_size);
+	}
+      det_pair_data.grow(fan_indices);
+      det_pair_data.fill(0);
+    }
 }
 
 void make_det_pair_data(DetPairData& det_pair_data,
 			const ProjData& proj_data,
 			const int segment_num,
 			const int ax_pos_num)
 {
-  make_det_pair_data(det_pair_data,
-		     *proj_data.get_proj_data_info_sptr(),
-		     segment_num,
-		     ax_pos_num);
-  const int num_detectors = 
-    det_pair_data.get_num_detectors();
-  const ProjDataInfoCylindricalNoArcCorr& proj_data_info =
-    dynamic_cast<const ProjDataInfoCylindricalNoArcCorr&>(*proj_data.get_proj_data_info_sptr());
-
-  shared_ptr<Sinogram<float> > 
-    pos_sino_ptr(new Sinogram<float>(proj_data.get_sinogram(ax_pos_num,segment_num)));
-  shared_ptr<Sinogram<float> > neg_sino_ptr;
-  if (segment_num == 0)
-    neg_sino_ptr = pos_sino_ptr;
+  if(proj_data.get_proj_data_info_sptr()->get_scanner_ptr()->get_scanner_geometry()== "BlocksOnCylindrical")
```

Comment:
so here we do need the cast, as we use `get_det_num_pair_for_view_tangential_pos_num`. Still, it's very ugly to have to duplicate the whole code. It's of course a consequence of our "faulty" hierarchy. What about creating a small function
```c++
void get_det_num_pair_for_view_tangential_pos_num(const ProjDataInfo proj_data_info&, ...)
{
  if (auto ptr = dynamic_cast<const ProjDataInfoCylindricalNoArcCorr const *>(&proj_data_info))
    ptr->get_det_num_pair_for_view_tangential_pos_num(...);
  else if (auto ptr = dynamic_cast<const ProjDataInfoGenericNoArcCorr const *>(&proj_data_info))
    ptr->get_det_num_pair_for_view_tangential_pos_num(...);
}
```
and just call that. No need for an `if` in the main code then. I think we use this trick elsewhere.